### PR TITLE
expression: refactor `JSON_QUOTE` / `JSON_UNQUOTE` 

### LINKED ISSUE #13501: builtin function `json_quote` execute panic
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

test 1

```
SELECT json_quote(rand());
```

test 2

```
CREATE TABLE t1(c1 int);
CREATE TABLE t2(c2 int);
INSERT INTO t1(`c1`) VALUES(1);
INSERT INTO t2(`c2`) VALUES(1);
SELECT json_quote(rand()) AS c3 FROM t1 JOIN t2;
```

2. What did you expect to see?

* MySQL 8.0

test 1

```
mysql> SELECT json_quote(rand());
ERROR 3064 (HY000): Incorrect type for argument 1 in function json_quote.
```

test  2
```
mysql> CREATE TABLE t1(c1 int);
Query OK, 0 rows affected (0.01 sec)

mysql> CREATE TABLE t2(c2 int);
Query OK, 0 rows affected (0.01 sec)

mysql> INSERT INTO t1(`c1`) VALUES(1);
Query OK, 1 row affected (0.00 sec)

mysql> INSERT INTO t2(`c2`) VALUES(1);
Query OK, 1 row affected (0.00 sec)

mysql> SELECT json_quote(rand()) AS c3 FROM t1 JOIN t2;
ERROR 3064 (HY000): Incorrect type for argument 1 in function json_quote.
```

3. What did you see instead?

test 1

```
mysql> SELECT json_quote(rand());
ERROR 2013 (HY000): Lost connection to MySQL server during query
```

test 1 logs

```
[2019/11/15 20:10:53.116 +08:00] [ERROR] [conn.go:619] ["connection running loop panic"] [conn=3749] [lastSQL="SELECT json_quote(rand())"] [err="runtime error: slice bounds out of range [:86173176] with capacity 15"] [stack="goroutine 20642773 [running]:
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0x26f6a40, 0xc00e3d79b0, 0xc008f46000)
github.com/pingcap/tidb@/server/conn.go:617 +0xee
panic(0x22bb400, 0xc00c0eb080)
runtime/panic.go:679 +0x1b2
github.com/pingcap/tidb/server.(*clientConn).writeResultset.func1(0x0, 0x2709260, 0xc00604e780, 0xc00f9313f0, 0x26f6a40, 0xc00e3d79b0, 0xc008f46000)
github.com/pingcap/tidb@/server/conn.go:1236 +0x597
panic(0x22bb400, 0xc00c0eb080)
runtime/panic.go:679 +0x1b2
github.com/pingcap/tidb/types/json.BinaryJSON.GetString(0xc00dd56e0b, 0xc034d1c4c1, 0x8, 0xf, 0xc02f8c3bc8, 0xc034d1c4b0, 0xc034d1c4c0)
github.com/pingcap/tidb@/types/json/binary.go:169 +0xb7
github.com/pingcap/tidb/types/json.BinaryJSON.Quote(0xc02f8c3b0b, 0xc034d1c4c1, 0x8, 0xf, 0xc02f8c3b90, 0x0)
github.com/pingcap/tidb@/types/json/binary_functions.go:60 +0x66
github.com/pingcap/tidb/expression.(*builtinJSONQuoteSig).vecEvalString(0xc0036b2540, 0xc00604e6e0, 0xc02f8c3b20, 0x0, 0x0)
github.com/pingcap/tidb@/expression/builtin_json_vec.go:317 +0x4e0
github.com/pingcap/tidb/expression.(*ScalarFunction).VecEvalString(0xc00604e690, 0x273ff60, 0xc010f41d00, 0xc00604e6e0, 0xc02f8c3b20, 0xc00f930d78, 0x1846ae0)
github.com/pingcap/tidb@/expression/scalar_function.go:56 +0x4b
github.com/pingcap/tidb/expression.evalOneVec(0x273ff60, 0xc010f41d00, 0x2741160, 0xc00604e690, 0xc00604e6e0, 0xc00604e820, 0x0, 0xc00bf81780, 0xc0044a5800)
github.com/pingcap/tidb@/expression/chunk_executor.go:142 +0xc81
github.com/pingcap/tidb/expression.(*defaultEvaluator).run(0xc00dd57480, 0x273ff60, 0xc010f41d00, 0xc00604e6e0, 0xc00604e820, 0x20300b, 0x2)
github.com/pingcap/tidb@/expression/evaluator.go:51 +0x3be
github.com/pingcap/tidb/expression.(*EvaluatorSuite).Run(0xc005f0ddf0, 0x273ff60, 0xc010f41d00, 0xc00604e6e0, 0xc00604e820, 0x0, 0x0)
github.com/pingcap/tidb@/expression/evaluator.go:123 +0xc9
github.com/pingcap/tidb/executor.(*ProjectionExec).unParallelExecute(0xc00d170700, 0x26f6a40, 0xc00e3d79b0, 0xc00604e820, 0x0, 0xc00f931000)
github.com/pingcap/tidb@/executor/projection.go:181 +0xfa
github.com/pingcap/tidb/executor.(*ProjectionExec).Next(0xc00d170700, 0x26f6a40, 0xc00e3d79b0, 0xc00604e820, 0xc00604e820, 0xc00f9310a8)
github.com/pingcap/tidb@/executor/projection.go:161 +0x73
github.com/pingcap/tidb/executor.Next(0x26f6a40, 0xc00e3d79b0, 0x26fda40, 0xc00d170700, 0xc00604e820, 0x0, 0x0)
github.com/pingcap/tidb@/executor/executor.go:210 +0x11f
github.com/pingcap/tidb/executor.(*recordSet).Next(0xc00604e730, 0x26f6a40, 0xc00e3d79b0, 0xc00604e820, 0xc00604e730, 0xc00604e820)
github.com/pingcap/tidb@/executor/adapter.go:117 +0x5a
github.com/pingcap/tidb/server.(*tidbResultSet).Next(0xc00604e780, 0x26f6a40, 0xc00e3d79b0, 0xc00604e820, 0x4, 0x400)
github.com/pingcap/tidb@/server/driver_tidb.go:368 +0x51
github.com/pingcap/tidb/server.(*clientConn).writeChunks(0xc008f46000, 0x26f6a40, 0xc00e3d79b0, 0x2709260, 0xc00604e780, 0xdc00, 0x22a5600, 0xc00604e780)
github.com/pingcap/tidb@/server/conn.go:1283 +0x364
github.com/pingcap/tidb/server.(*clientConn).writeResultset(0xc008f46000, 0x26f6a40, 0xc00e3d79b0, 0x2709260, 0xc00604e780, 0xc00000de00, 0x0, 0x0, 0x0)
github.com/pingcap/tidb@/server/conn.go:1249 +0x206
github.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc008f46000, 0x26f6a40, 0xc00e3d79b0, 0xc00c0eaf81, 0x19, 0x0, 0x0)
github.com/pingcap/tidb@/server/conn.go:1166 +0x209
github.com/pingcap/tidb/server.(*clientConn).dispatch(0xc008f46000, 0x26f6a40, 0xc00e3d79b0, 0xc00c0eaf81, 0x1a, 0x19, 0x0, 0x0)
github.com/pingcap/tidb@/server/conn.go:850 +0x5b3
github.com/pingcap/tidb/server.(*clientConn).Run(0xc008f46000, 0x26f6a40, 0xc00e3d79b0)
github.com/pingcap/tidb@/server/conn.go:672 +0x273
github.com/pingcap/tidb/server.(*Server).onConn(0xc000e9c480, 0xc008f46000)
github.com/pingcap/tidb@/server/server.go:438 +0x49f
created by github.com/pingcap/tidb/server.(*Server).Run
github.com/pingcap/tidb@/server/server.go:355 +0x759
"]
```

test 2

```
mysql> CREATE TABLE t1(c1 int);
Query OK, 0 rows affected (1.51 sec)

mysql> CREATE TABLE t2(c2 int);
Query OK, 0 rows affected (0.51 sec)

mysql> INSERT INTO t1(`c1`) VALUES(1);
Query OK, 1 row affected (0.01 sec)

mysql> INSERT INTO t2(`c2`) VALUES(1);
Query OK, 1 row affected (0.01 sec)

mysql> SELECT json_quote(rand()) AS c3 FROM t1 JOIN t2;
ERROR 1105 (HY000): runtime error: slice bounds out of range [:1505392] with capacity 15
```

test 2 logs

```
[2019/11/15 20:08:46.494 +08:00] [ERROR] [projection.go:378] ["projection executor panicked"] [error="runtime error: slice bounds out of range [:1505392] with capacity 15"] [stack="goroutine 20681980 [running]:
github.com/pingcap/tidb/util.GetStack(...)
github.com/pingcap/tidb@/util/misc.go:60
github.com/pingcap/tidb/executor.recoveryProjection(0xc0060ceab0, 0x22bb400, 0xc00d5a9de0)
github.com/pingcap/tidb@/executor/projection.go:377 +0x90
github.com/pingcap/tidb/executor.(*projectionWorker).run.func1(0xc0009fcf90)
github.com/pingcap/tidb@/executor/projection.go:348 +0x4e
panic(0x22bb400, 0xc00d5a9de0)
runtime/panic.go:679 +0x1b2
github.com/pingcap/tidb/types/json.BinaryJSON.GetString(0xc00938700b, 0xc02e4aaa81, 0x8, 0xf, 0xc016f5e508, 0xc02e4aaa70, 0xc02e4aaa80)
github.com/pingcap/tidb@/types/json/binary.go:169 +0xb7
github.com/pingcap/tidb/types/json.BinaryJSON.Quote(0xc00e75dd0b, 0xc02e4aaa81, 0x8, 0xf, 0xc016f5e4d0, 0x0)
github.com/pingcap/tidb@/types/json/binary_functions.go:60 +0x66
github.com/pingcap/tidb/expression.(*builtinJSONQuoteSig).vecEvalString(0xc00d076060, 0xc0594a5d60, 0xc00e75ddc0, 0x0, 0x0)
github.com/pingcap/tidb@/expression/builtin_json_vec.go:317 +0x4e0
github.com/pingcap/tidb/expression.(*ScalarFunction).VecEvalString(0xc002980280, 0x273ff60, 0xc010f41d00, 0xc0594a5d60, 0xc00e75ddc0, 0xc00f8504c8, 0x1846ae0)
github.com/pingcap/tidb@/expression/scalar_function.go:56 +0x4b
github.com/pingcap/tidb/expression.evalOneVec(0x273ff60, 0xc010f41d00, 0x2741160, 0xc002980280, 0xc0594a5d60, 0xc0594a5e50, 0x0, 0xc00f850708, 0x1e08c83)
github.com/pingcap/tidb@/expression/chunk_executor.go:142 +0xc81
github.com/pingcap/tidb/expression.(*defaultEvaluator).run(0xc00e6a9880, 0x273ff60, 0xc010f41d00, 0xc0594a5d60, 0xc0594a5e50, 0x0, 0x0)
github.com/pingcap/tidb@/expression/evaluator.go:51 +0x3be
github.com/pingcap/tidb/expression.(*EvaluatorSuite).Run(0xc0060ce9a0, 0x273ff60, 0xc010f41d00, 0xc0594a5d60, 0xc0594a5e50, 0xc00f850780, 0xd234cf)
github.com/pingcap/tidb@/expression/evaluator.go:123 +0xc9
github.com/pingcap/tidb/executor.(*projectionWorker).run(0xc00e6a9a40, 0x26f6a40, 0xc00e3d79b0)
github.com/pingcap/tidb@/executor/projection.go:362 +0x143
created by github.com/pingcap/tidb/executor.(*ProjectionExec).prepare
github.com/pingcap/tidb@/executor/projection.go:246 +0x707
"]
[2019/11/15 20:08:46.494 +08:00] [WARN] [conn.go:688] ["command dispatched failed"] [conn=3749] [connInfo="id:3749, addr:172.17.0.2:56420 status:10, collation:utf8_general_ci, user:root"] [command=Query] [status="inTxn:0, autocommit:1"] [sql="SELECT json_quote(rand()) AS c3 FROM t1 JOIN t2"] [err="runtime error: slice bounds out of range [:1505392] with capacity 15
github.com/pingcap/tidb/executor.recoveryProjection
github.com/pingcap/tidb@/executor/projection.go:375
github.com/pingcap/tidb/executor.(*projectionWorker).run.func1
github.com/pingcap/tidb@/executor/projection.go:348
runtime.gopanic
runtime/panic.go:679
runtime.goPanicSliceAcap
runtime/panic.go:93
github.com/pingcap/tidb/types/json.BinaryJSON.GetString
github.com/pingcap/tidb@/types/json/binary.go:169
github.com/pingcap/tidb/types/json.BinaryJSON.Quote
github.com/pingcap/tidb@/types/json/binary_functions.go:60
github.com/pingcap/tidb/expression.(*builtinJSONQuoteSig).vecEvalString
github.com/pingcap/tidb@/expression/builtin_json_vec.go:317
github.com/pingcap/tidb/expression.(*ScalarFunction).VecEvalString
github.com/pingcap/tidb@/expression/scalar_function.go:56
github.com/pingcap/tidb/expression.evalOneVec
github.com/pingcap/tidb@/expression/chunk_executor.go:142
github.com/pingcap/tidb/expression.(*defaultEvaluator).run
github.com/pingcap/tidb@/expression/evaluator.go:51
github.com/pingcap/tidb/expression.(*EvaluatorSuite).Run
github.com/pingcap/tidb@/expression/evaluator.go:123
github.com/pingcap/tidb/executor.(*projectionWorker).run
github.com/pingcap/tidb@/executor/projection.go:362
runtime.goexit
runtime/asm_amd64.s:1357"]
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
Release Version: v4.0.0-alpha-857-gf94eac07d
Git Commit Hash: f94eac07df296ac3c1899c2b77b1ede8a71c7936
Git Branch: master
UTC Build Time: 2019-11-14 03:00:55
GoVersion: go version go1.13 linux/amd64
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by Deardrops**
### What problem does this PR solve? 

Fix #13501 
### What is changed and how it works?

We change the parameter type of `JSON_QUOTE` and `JSON_UNQUOTE` from `JSON` to `string`.
Checking the parameter type in `getFunction`.
Meanwhile, Adding lots of tests.

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported variable/fields change

Side effects

 - None

Related changes

 - None

Release note

 - None

### CODE REVIEW

File: expression/builtin_json.go [hunk 1]
```diff
@@ -1038,6 +1038,8 @@ func (b *builtinJSONQuoteSig) evalString(row chunk.Row) (res string, isNull bool
 	j, isNull, err = b.args[0].EvalJSON(b.ctx, row)
```
Comment (SunRunAway) [inline expression/builtin_json.go]: Is it better to call EvalString here?
And we should check the type of the argument in `getFunction`.

**Comment by Deardrops** [comment]
@SunRunAway comment addressed, PTAL

File: expression/builtin_json.go [hunk 2]
```diff
@@ -186,22 +187,39 @@ func (c *jsonUnquoteFunctionClass) getFunction(ctx sessionctx.Context, args []Ex
 	if err := c.verifyArgs(args); err != nil {
 		return nil, err
 	}
-	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETString, types.ETJson)
+	switch args[0].GetType().Tp {
+	case mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeString, mysql.TypeNull, mysql.TypeJSON,
```
Comment (wjhuang2016) [inline expression/builtin_json.go]: How about put it into a list and add a function to check them?

File: expression/builtin_json.go [hunk 3]
```diff
@@ -186,22 +187,39 @@ func (c *jsonUnquoteFunctionClass) getFunction(ctx sessionctx.Context, args []Ex
...
-func (b *builtinJSONUnquoteSig) evalString(row chunk.Row) (res string, isNull bool, err error) {
-	var j json.BinaryJSON
-	j, isNull, err = b.args[0].EvalJSON(b.ctx, row)
-	if isNull || err != nil {
+func (b *builtinJSONUnquoteSig) evalString(row chunk.Row) (string, bool, error) {
+	str, isNull, err := b.args[0].EvalString(b.ctx, row)
+	if err != nil {
 		return "", isNull, err
+	} else if isNull {
+		return "NULL", true, err
```
Comment (wjhuang2016) [inline expression/builtin_json.go]: Why return "NULL" here? Is it necessary?

File: expression/builtin_json_vec.go [hunk 4]
```diff
@@ -826,11 +828,19 @@ func (b *builtinJSONUnquoteSig) vecEvalString(input *chunk.Chunk, result *chunk.
 			result.AppendNull()
 			continue
 		}
-		res, err := buf.GetJSON(i).Unquote()
-		if err != nil {
-			return err
+		str := buf.GetString(i)
+		tlen := len(str)
+		if tlen >= 2 {
```
Comment (wjhuang2016) [inline expression/builtin_json_vec.go]: Same logic with previous code? Use a function to do that.

File: expression/builtin_json.go [hunk 5]
```diff
@@ -186,22 +187,39 @@ func (c *jsonUnquoteFunctionClass) getFunction(ctx sessionctx.Context, args []Ex
...
+	}
+	tlen := len(str)
```
Comment (AilinKid) [inline expression/builtin_json.go]: this code is quite same as `json.Unquote()`,
I think we should keep `arg[0].EvalJSON` & `json.Unquote()`, 
just add a check for `args[0].EvalString` at the beginning is ok.

File: expression/builtin_json_vec.go [hunk 6]
```diff
@@ -826,11 +828,19 @@ func (b *builtinJSONUnquoteSig) vecEvalString(input *chunk.Chunk, result *chunk.
...
+			head, tail := str[0], str[tlen-1]
+			if head == '"' && tail == '"' {
+				// Remove prefix and suffix '"' before unquoting
+				str, err = json.UnquoteString(str[1 : tlen-1])
```
Comment (AilinKid) [inline expression/builtin_json_vec.go]: ditto here.

File: expression/builtin_json.go [hunk 7]
```diff
@@ -186,22 +187,39 @@ func (c *jsonUnquoteFunctionClass) getFunction(ctx sessionctx.Context, args []Ex
...
+	}
+}
+
+func (b *builtinJSONUnquoteSig) evalString(row chunk.Row) (string, bool, error) {
+	str, isNull, err := b.args[0].EvalString(b.ctx, row)
+	if err != nil {
 		return "", isNull, err
+	} else if isNull {
+		return "", true, err
+	}
```
Comment (SunRunAway) [inline expression/builtin_json.go]:
```suggestion
	if isNull || err != nil {
        return "", isNull, err
    }
```

File: expression/builtin_json.go [hunk 8]
```diff
@@ -1026,20 +1044,34 @@ func (c *jsonQuoteFunctionClass) getFunction(ctx sessionctx.Context, args []Expr
...
+}
+
+func (b *builtinJSONQuoteSig) evalString(row chunk.Row) (string, bool, error) {
+	str, isNull, err := b.args[0].EvalString(b.ctx, row)
+	if err != nil {
 		return "", isNull, err
+	} else if isNull {
+		return "", true, err
 	}
```
Comment (SunRunAway) [inline expression/builtin_json.go]: ditto.

**Comment by Deardrops** [comment]
@SunRunAway Comment addressed, PTAL

**Pull request merged**